toolkit/tools/pkg/imagecustomizerlib/defaultgrubutils.go

```diff
@@ -0,0 +1,256 @@
+) (grubDefaultVarAssign, []grubConfigLinuxArg, int, error) {
+	varAssigns, err := findGrubDefaultVarAssigns(defaultGrubContent)
+	if err != nil {
+		err = fmt.Errorf("failed to parse %s file:\n%w", installutils.GrubDefFile, err)
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	// Find the variable's (e.g. GRUB_CMDLINE_LINUX) line.
+	cmdLineVarAssign, err := findGrubDefaultVarAssign(varAssigns, varName)
+	if err != nil {
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	// The (parsed) string value of the variable is copied verbatim into the grub.cfg file.
+	// So, args can have quotes if needed, but those quotes will be double escaped.
+	argsString := cmdLineVarAssign.Value
+	grubTokens, err := grub.TokenizeConfig(argsString)
+	if err != nil {
+		err = fmt.Errorf("failed to parse %s's value:\n%w", varName, err)
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	var insertAt int
+	if varName == defaultGrubCmdlineLinuxDefault {
+		// GRUB_CMDLINE_LINUX_DEFAULT variable has the $kernelopts arg.
+		// Any args inserted should be inserted before $kernelopts.
+		insertAt, err = findCommandLineInsertAt(grubTokens)
+		if err != nil {
+			err = fmt.Errorf("failed to parse %s's value args:\n%w", varName, err)
+			return grubDefaultVarAssign{}, nil, 0, err
+		}
+	} else {
+		// Insert args at the end of the string.
+		insertAt = len(argsString)
+	}
+
+	args, err := parseCommandLineArgs(grubTokens)
+	if err != nil {
+		err = fmt.Errorf("failed to parse %s's value args:\n%w", varName, err)
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	return cmdLineVarAssign, args, insertAt, nil
+}
+
+func addExtraCommandLineToDefaultGrubFile(defaultGrubContent string, extraCommandLine string) (string, error) {
+	cmdLineVarAssign, _, insertAt, err := getGrubDefaultLinuxArgs(defaultGrubContent,
+		defaultGrubCmdlineLinuxDefault)
+	if err != nil {
+		return "", err
+	}
+
+	argsString := cmdLineVarAssign.Value
+
+	// Add the extra command-line args.
+	argsString = argsString[:insertAt] + " " + extraCommandLine + " " + argsString[insertAt:]
+
+	// Rewrite GRUB_CMDLINE_LINUX_DEFAULT line.
+	defaultGrubContent = rewriteDefaultGrubLine(defaultGrubContent, cmdLineVarAssign, argsString)
+	return defaultGrubContent, nil
+}
+
+func updateDefaultGrubKernelCommandLineArgs(defaultGrubContent string, varName string, argsToRemove []string,
+	newArgs []string,
+) (string, error) {
+	cmdLineVarAssign, args, insertAt, err := getGrubDefaultLinuxArgs(defaultGrubContent, varName)
+	if err != nil {
+		return "", err
+	}
+
+	value := cmdLineVarAssign.Value
+	value, err = updateKernelCommandLineArgsHelper(value, args, insertAt, argsToRemove, newArgs)
+	if err != nil {
+		return "", err
+	}
+
+	// Rewrite GRUB_CMDLINE_LINUX line.
+	defaultGrubContent = rewriteDefaultGrubLine(defaultGrubContent, cmdLineVarAssign, value)
+	return defaultGrubContent, nil
+}
+
+func rewriteDefaultGrubLine(defaultGrubContent string, varAssign grubDefaultVarAssign, newValue string) string {
```

Comment:
would it make sense to rename this to replace token? The name is a bit hard to understand